lib/network.go

```diff
@@ -2351,11 +2352,101 @@ func (msg *MsgDeSoHeader) String() string {
 // ==================================================================
 
 type BlockProducerInfo struct {
+	Version BlockProducerInfoVersion
+
+	// ECDSA public key for the block producer.
 	PublicKey []byte
+	// The block producer's ECDSA signature for the block. This field is used in
+	// BlockProducerInfo version 0, and is deprecated from version 1 onwards.
 	Signature *btcec.Signature
+
+	// The BLS public key of the validator who constructed this block. This field is
+	// populated starting in BlockProducerInfo version 1.
+	VotingPublicKey *bls.PublicKey
+	// The validator's partial BLS signature of the (ProposedInView, BlockHash) pair
+	// for enclosing block. This signature proves the validator proposed the block,
+	// and also acts as the validator's vote for this block. This filed is only populated
+	// starting in BlockProducerInfo version 1.
+	VotePartialSignature *bls.Signature
+}
+
+// Byte encoder for the BlockProducerInfo with support for versioning. The encoder only
+// supports BlockProducerInfo version 1 and above. For the legacy version 0, use the
+// BlockProducerInfo.Serialize_Legacy() method instead.
+func (bpi *BlockProducerInfo) ToBytes() ([]byte, error) {
+	// BlockProducerInfo version 0 is not supported.
+	if bpi.Version == BlockProducerInfoVersion0 {
+		return nil, fmt.Errorf("BlockProducerInfo.ToBytes: BlockProducerInfo version 0 not supported")
+	}
+
+	encodedBytes := []byte{}
+
+	// Required Version field
+	encodedBytes = append(encodedBytes, bpi.Version)
+
+	// Required ECDSA PublicKey
+	if len(bpi.PublicKey) == 0 {
+		return nil, fmt.Errorf("BlockProducerInfo.ToBytes: PublicKey is required")
+	}
+	encodedBytes = append(encodedBytes, EncodeByteArray(bpi.PublicKey)...)
```

Comment:
I think all public keys should have the same length. If so, we can just concat the bytes instead of prepending w/ the length of the byte array.

Reply:
Sounds good. We'll update the fields type the the strongly typed `PublicKey` so it can reuse the type byte encoder.